Fix PyTorch tensor handling in `CupyOps.asarray`

shadeMe opened the pull request:
This PR fixes the following and adds a regression test:

- The original invocation of `cupy.cuda.MemoryPointer` was missing the mandatory `offset` argument. This should be `0`.
- `cupy.cuda.MemoryPointer` expects a wrapper to a `cupy`-owned CUDA memory and not a raw pointer to it. So, CUDA memory allocated by external libraries needs to be wrapped with `UnownedMemory` and then passed to the array c'tor [[1](https://github.com/cupy/cupy/issues/4644#issuecomment-775480279)].
- We weren't testing that the incoming array-like obj was owned by a CUDA device.
- We were incorrectly passing the array stride as its shape.

Should fix #564.

shadeMe commented:
@explosion-bot please test_gpu

shadeMe commented:
One more time with feeling: @explosion-bot please test_gpu

shadeMe commented:
@explosion-bot please test_gpu

danieldk commented on thinc/backends/cupy_ops.py:
```diff
@@ -75,11 +75,23 @@ def asarray(self, data, dtype=None):
-            pointer = cupy.cuda.MemoryPointer(data.data_ptr())
-            shape = data.stride()
-            array = self.xp.ndarray(shape, memptr=pointer, **dtype)
+            assert data.device.type == "cuda", "Non-CUDA tensor passed to CupyOps"
```
Since this method is part of the public API, it's probably better to throw a `ValueError` here?

danieldk commented on thinc/backends/cupy_ops.py:
```diff
@@ -75,11 +75,23 @@ def asarray(self, data, dtype=None):
+            )
+
+            pointer = cupy.cuda.MemoryPointer(ext_pointer_wrapper, offset=0)
+            array = self.xp.ndarray(shape=data.shape, memptr=pointer, **dtype)
```
The semantics of `dtype` are incorrect. If `dtype` is not specified, the data type of the original array must be used, otherwise:

```
>>> x = torch.tensor([42.0], device='cuda')
>>> y = ops.asarray(x)
>>> y
array([5.48372203e-315])
```

If `dtype` is specified and differs from the `dtype` of the original array, a copy should be made (converting to the specified `dtype`).

shadeMe commented:
Added support for zero-alloc conversions for Torch, TF and MXNet arrays and removed the error for CPU-allocated inputs - I think we'll want to just copy it to the GPU in such cases instead of raising a hard error.

shadeMe commented:
@explosion-bot please test_gpu

adrianeboyd commented on thinc/backends/cupy_ops.py:
```diff
@@ -19,6 +19,14 @@
 from .numpy_ops import NumpyOps
 from . import _custom_kernels
 from ..types import DeviceTypes
+from ..util import (
+    torch2xp,
+    tensorflow2xp,
+    is_mxnet_array,
+    mxnet2xp,
+    is_torch_array,
+    is_tensorflow_array,
+)
```
We usually reformat (manually) with multiple imports rather than using this kind of multi-line import:

```suggestion
from ..util import torch2xp, tensorflow2xp, is_mxnet_array, mxnet2xp
from ..util import is_torch_array, is_tensorflow_array
```

svlandeg reviewed (commented):
Looks good to me, but will defer the final review & merging to @danieldk if he's happy :-)

danieldk reviewed (commented):
Looks good to me, though I'll run my 'test all the PyTorch versions' script just to be sure.
